src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -1457,6 +1462,10 @@ protected virtual void Dispose(bool disposing)
                     catch { } // If it fails, simply give up.
                     m_eventSourceEnabled = false;
                 }
+
+                // Wait till active writes have finished (stop waiting at 1 second)
+                SpinWait.SpinUntil(() => m_activeWritesCount == 0, 1000);
```

Comment:
How do we know 1 second is the right value to wait?

Reply:
This should have a big old `TODO` over it saying "until I find evidence for a better value".

I'm not even sure the timeout is appropriate. My fear is there is a code path after a call to one of the core `Write` methods that would take a lock and having this infinitely loop would be asking for a deadlock of some kind.

Reply:
Let me make a better suggestion here. When I was writing the ETW collector provided by VS this problem was endemic to all ETW collector designs. A solution we came up with was during shutdown we inserted an event into the ETW stream. When we observed that event we knew we were "synchronized" with the request. This is entirely true for reasons I can explain offline but it is close enough that VS's ETW collector has never had missed events due to the logic. Timeouts like this in ETW should be avoided at all costs. One second actually is a very reasonable value though &ndash; I can also explain that offline if interested.